### DISCUSSION TITLE
Porting HOST algorithm from Isaac Gym to Isaac Lab

### DISCUSSION THREAD

**Discussion by francescacraievich** [community/community#164800] [Other Feature Feedback, Questions, & Ideas]
I’m working with the team on porting a custom reinforcement learning algorithm from Isaac Gym (Legged-Gym) to Isaac Lab using Isaac Orbit, and I’d really appreciate any advice or guidance from the community.

--------
The original implementation is based on the paper: 

**"Learning Humanoid Standing-Up Control Across Diverse Postures"** by Tao Huang and partners.

The code is built upon Nvidia’s Legged-Gym library (using Isaac Gym), and defines a multi-stage standing-up behavior for a humanoid robot. The agent is trained with PPO and leverages custom design features like:

 - Multi-critic reward optimization, grouped by task stages (righting, kneeling, rising)
 - Curriculum learning, with a vertical force applied at the start of training
 - Action rescaler β to control joint speed/torque smoothly
 - Smoothness regularization to reduce oscillatory motion
 - Domain randomization for sim2real transfer and terrain diversity
-----

I want to recreate the same learning environment and behavior inside Isaac Lab, using the Orbit framework. Specifically:
What I'm looking for:

 1. How can I implement a multi-critic RL setup or simulate one using Orbit's task and reward structures?
 2. Any recommendations for building custom curriculum learning (applying force, changing difficulty)?
 3. Best practices to add a PD controller with action rescaling β to a humanoid robot in Isaac Orbit?
 4. How to separate and log multiple reward components (for better interpretability / debugging)?
 5. Examples of domain randomization and initial posture variation using Orbit’s Scene + Randomization API?
 6. Are there existing examples or repos that implement something similar with humanoids in Isaac Lab?

-------
If you’ve worked on similar problems or have seen relevant examples, I’d love to hear from you. Thanks in advance for your time and any suggestions 🙏

Best regards,

Francesca

**Comment by dannz510** [comment]
You're tackling an ambitious and interesting project, Francesca\! Porting a complex RL algorithm like "Learning Humanoid Standing-Up Control Across Diverse Postures" from Legged-Gym (Isaac Gym) to Isaac Lab (Isaac Orbit) is a significant undertaking, but Isaac Lab and Orbit are designed to be quite flexible.

Let's break down your questions and provide guidance based on Orbit's architecture and common RL practices.

### General Advice for Transitioning from Legged-Gym to Isaac Lab/Orbit

  * **Understand Orbit's Abstractions:** Orbit introduces concepts like `Scene`, `Robot`, `AssetBase`, `RewardFunctions`, `Terminations`, `CurriculumMgr`, and the `RLTask` structure. Familiarize yourself with these, as they are the building blocks for your environment.
  * **Modular Design:** Orbit encourages modularity. Think about how each of your custom features (multi-critic, curriculum, action rescaler, smoothness) can be encapsulated into separate components or functions that plug into the Orbit framework.
  * **Leverage Existing Examples:** Isaac Lab comes with many examples. Even if they don't exactly match your humanoid standing-up task, studying how they implement robots, rewards, terminations, and randomization will be incredibly helpful. Pay close attention to the `omni.isaac.orbit.tasks` and `omni.isaac.orbit.robots` modules.

-----

### Addressing Your Specific Questions:

#### 1\. How to implement a multi-critic RL setup or simulate one using Orbit's task and reward structures?

Orbit's standard `RLTask` framework is designed for a single scalar reward returned by the environment. A true multi-critic setup (where different critics optimize different reward functions independently, feeding into a single actor) isn't directly exposed as a plug-and-play feature within Orbit's `RLTask`.

However, you can **simulate** or integrate this concept:

  * **Combined Reward Function with Weighted Sum:**

      * **Orbit's approach:** The most common way in Orbit is to define a single `RewardFunctions` class within your `RLTask`.
      * **Implementation:** Inside this class, you'll compute all your individual reward components (e.g., `righting_reward`, `kneeling_reward`, `rising_reward`, `smoothness_reward`).
      * **Weighting:** You then combine these into a single scalar reward using weights. These weights can be static, or, more relevant to your multi-stage approach, they can be **dynamic** based on the current stage of the standing-up task.
      * **Stage-dependent weighting:** You'll need to define a state variable or a set of conditions that determine which stage the robot is in (e.g., based on center of mass height, joint angles, or specific contact points). Based on the current stage, you would adjust the weights of your individual reward components, effectively prioritizing certain behaviors. This isn't a "multi-critic" in the literal sense, but it achieves a similar *behavioral shaping* goal.

  * **Using a Custom RL Algorithm Outside Orbit (Advanced):**

      * If you *truly* need separate critic networks, you would typically need to integrate Orbit with a more flexible RL library that supports custom policy architectures (e.g., Stable Baselines3's PPO with custom networks, or your own PyTorch/TensorFlow implementation).
      * In this scenario, Orbit would serve as the environment (`gym.Env` interface), and your external RL script would handle the multi-critic network architecture, training loops, and data collection. Orbit would still provide the individual reward components, and your external script would feed these into the respective critics. This is more involved and might be a later optimization if the weighted sum approach isn't sufficient.

#### 2\. Any recommendations for building custom curriculum learning (applying force, changing difficulty)?

Orbit provides a `CurriculumMgr` within the `RLTask` framework, which is perfect for this.

  * **`CurriculumMgr`:**

      * You define specific curriculum types (e.g., `RewardCurriculum`, `ProgressCurriculum`).
      * You link these curricula to variables in your task definition (e.g., initial force magnitude, terrain difficulty parameters, target pose).
      * The `CurriculumMgr` handles updating these variables based on episode success, average reward, or a fixed schedule.

  * **Applying Force at Start of Training:**

      * You can add a `FixedImpulse` or `ConstantForce` component to your robot or specific parts of the scene within your `RLTask`.
      * This force's magnitude can be a curriculum variable.
      * Initially, set this variable to a high value. As training progresses and the robot learns, decrease the force magnitude (e.g., using a `ProgressCurriculum` tied to `env.episode_length` or `env.reward`).

  * **Changing Difficulty:**

      * **Initial Posture:** Randomize initial postures, and over curriculum stages, perhaps narrow the range of randomization or guide it towards more challenging starting points.
      * **Terrain Diversity (via Domain Randomization):** Use Orbit's randomization capabilities (discussed below) to introduce more complex terrains over time. This can be tied to a curriculum variable.
      * **Target Pose:** Gradually increase the complexity of the target standing-up posture (if applicable).
      * **Disturbances:** Introduce external disturbances (e.g., random pushes or torques) with increasing magnitude or frequency as the curriculum progresses.

#### 3\. Best practices to add a PD controller with action rescaling $\\beta$ to a humanoid robot in Isaac Orbit?

Orbit's `Robot` class and `ActuatorGroup` system are designed for this.

  * **PD Controller Setup:**

      * When defining your robot (e.g., using `RobotCfg`), you specify the actuator type for each joint (e.g., `ActuatorType.TORQUE` for direct torque control, or `ActuatorType.POSITION` if you want Orbit to handle the low-level PD for target positions).
      * For a custom PD controller that you want to apply before sending commands, you'd typically send position or velocity targets to Orbit's actuators, and then Orbit's internal PD controller would convert these to torques.
      * If you need finer control and want to implement your *own* PD loop *before* sending torques, you would define your actuators as `ActuatorType.TORQUE`. Then, in your `RLTask`'s `_apply_action()` method, you would calculate the desired torques based on your PD gains and the action from the policy, and then use `robot.set_joint_efforts()`.

  * **Action Rescaling $\\beta$:**

      * The action rescaler $\\beta$ (which often controls joint speed/torque limits) can be integrated during the `_apply_action()` phase of your `RLTask`.
      * **Policy Output:** Your policy will output a raw action (e.g., desired joint positions or velocities, or direct torques).
      * **Scaling:** Before applying these actions to the robot, you would perform the rescaling:
        ```python
        # Assuming `actions` are from your policy, e.g., normalized to [-1, 1]
        # and `beta` is your rescaling factor (e.g., max_speed or max_torque)

        # Example: Scaling for joint velocity targets
        desired_joint_velocities = actions * beta_velocity_scale

        # Example: Scaling for joint torque commands
        desired_joint_torques = actions * beta_torque_scale

        # Then apply to robot:
        self.robot.set_joint_velocity_targets(desired_joint_velocities)
        # OR
        self.robot.set_joint_efforts(desired_joint_torques)
        ```
      * **Making $\\beta$ Learnable/Curriculum-based:** You could potentially make $\\beta$ a parameter that is also part of your policy's output (if it's a learnable component) or, more likely given the "control joint speed/torque smoothly" description, a curriculum-controlled parameter that changes over training stages.

#### 4\. How to separate and log multiple reward components (for better interpretability / debugging)?

Orbit's `RewardFunctions` class and the logging capabilities of `warp_track` or external logging tools are key here.

  * **Within `RewardFunctions`:**

      * In your `RewardFunctions` class (e.g., `MyHumanoidRewards`), define separate methods for each reward component:
        ```python
        from omni.isaac.orbit.managers import RewardFunctionsCfg
        from omni.isaac.orbit.managers.reward_manager import RewardTerm

        @RewardFunctionsCfg.register("MyHumanoidRewards")
        class MyHumanoidRewards(RewardTerm):
            def __init__(self, cfg: RewardFunctionsCfg, env: RLTask):
                super().__init__(cfg, env)
                # Initialize any necessary references to robot or scene
                self.robot = env.robot

            def righting_reward(self, **kwargs):
                # Calculate reward based on robot orientation or COM
                # ...
                return reward_value

            def kneeling_reward(self, **kwargs):
                # Calculate reward based on specific joint angles for kneeling
                # ...
                return reward_value

            def total_reward(self, **kwargs):
                # Combine individual rewards with weights
                total_rew = (
                    self.righting_reward(**kwargs) * self.cfg.righting_weight +
                    self.kneeling_reward(**kwargs) * self.cfg.kneeling_weight +
                    # ... other rewards
                )
                return total_rew
        ```
      * **`RewardFunctionsCfg`:** In your `RLTaskCfg`, you'll define your `RewardFunctionsCfg` and specify the individual terms and their weights.
        ```python
        class MyHumanoidRLTaskCfg(RLTaskCfg):
            class RewardsCfg(RewardFunctionsCfg):
                # Define specific reward terms and their weights
                righting_reward = {"func": MyHumanoidRewards.righting_reward, "weight": 1.0}
                kneeling_reward = {"func": MyHumanoidRewards.kneeling_reward, "weight": 0.5}
                # ... other terms
                # This is crucial for logging individual components:
                # The total_reward will be the one returned to the RL agent
                total_reward_sum = {"func": MyHumanoidRewards.total_reward, "weight": 1.0}
        ```

  * **Logging:**

      * **WandB/TensorBoard with Orbit's Logger:** Orbit integrates with `warp_track` which can be configured to log to TensorBoard or Weights & Biases (WandB).
      * **`extras` Dictionary:** When you compute your reward terms in the `RewardFunctions` class, you can add them to the `extras` dictionary that `RLTask` passes to the logger.
        ```python
        # Inside your _update_metrics method or similar in RLTask
        # Or even better, define in your RewardFunctions class if it's responsible for computing them
        extras = {
            "metrics/reward_righting": righting_rew.mean(),
            "metrics/reward_kneeling": kneeling_rew.mean(),
            "metrics/reward_smoothness": smoothness_rew.mean(),
            # ... and so on
        }
        self.common_step_data["extras"] = extras
        ```
      * The `RLTask`'s `step()` method collects these `extras` and logs them. This allows you to see the individual contributions to the total reward over time in your logging dashboard (TensorBoard/WandB).

#### 5\. Examples of domain randomization and initial posture variation using Orbit’s Scene + Randomization API?

Orbit's `SceneEntityCfg` and `RandomizationCfg` are built for this.

  * **Scene Randomization:**

      * In your `RLTaskCfg`, you'll define `SceneEntityCfg` for various assets (robot, terrain, props).
      * Within `RandomizationCfg`, you can specify:
          * **Initial State Randomization:** For your humanoid, you'd randomize the initial joint positions and velocities. You can specify ranges for each joint.
            ```python
            from omni.isaac.orbit.scene import InteractiveSceneCfg
            from omni.isaac.orbit.terrains import TerrainImporterCfg
            from omni.isaac.orbit.randomization import RandomizationCfg

            class MyHumanoidRLTaskCfg(RLTaskCfg):
                class SceneCfg(InteractiveSceneCfg):
                    robot = RobotCfg(
                        # ... other robot configs
                        init_state=RobotCfg.InitialStateCfg(
                            pos=(0.0, 0.0, 1.0), # Example initial position
                            # Randomize joint positions from min_angles to max_angles
                            joint_pos=InitialJointPosesCfg(
                                joints={"joint_name_1": (min_angle_1, max_angle_1),
                                        "joint_name_2": (min_angle_2, max_angle_2)},
                                default_joint_pos=[0.0] * num_joints # if not specified individually
                            ),
                            # Randomize joint velocities
                            joint_vel=(min_vel, max_vel)
                        )
                    )
                    # Example: Add a ground plane
                    ground = GroundPlaneCfg()

                    # Example: Add terrain randomization
                    terrain = TerrainImporterCfg(
                        # ... terrain parameters
                        # This allows randomizing terrain features like roughness, slope, etc.
                    )

                # Randomization manager
                randomization = RandomizationCfg(
                    # Add randomizations that apply to the scene or robot
                    # e.g., randomize external forces, mass properties, friction
                    # You can define custom randomizers here too
                    randomizers=[
                        # Add a randomizer for external forces applied to robot
                        {"func": "apply_external_force_on_robot", "params": {...}},
                        # Add a randomizer for friction coefficients
                        {"func": "randomize_friction", "params": {...}},
                        # You can also define custom randomization functions in your task
                        {"func": "randomize_robot_mass", "params": {...}},
                    ]
                )
            ```
          * **Physical Properties:** Randomize robot mass, friction coefficients of its parts, damping, etc.
          * **External Disturbances:** Apply random forces or torques to the robot.
          * **Lighting/Visuals:** Randomize lighting, textures for sim2real.

  * **Terrain Diversity:**

      * Use `TerrainImporterCfg` to define a diverse set of terrains (e.g., flat, rough, stairs, ramps). Orbit can then randomly sample these terrains for each episode or based on a curriculum.

#### 6\. Are there existing examples or repos that implement something similar with humanoids in Isaac Lab?

While there isn't a direct "standing-up humanoid" example *exactly* like yours, here are highly relevant examples within Isaac Lab/Orbit that you should dissect:

  * **`omni.isaac.orbit_envs.locomotion.legged_robot` (Specifically `legged_robot_cfg.py` and `legged_robot_rewards.py`):**
      * This is the most relevant starting point. It implements a general quadrupedal locomotion task.
      * **Key Learnings:**
          * How to define a `RobotCfg` for a legged robot (even if not a humanoid).
          * How to implement various reward components (e.g., `stand_still_reward`, `track_lin_vel_reward`, `track_ang_vel_reward`, `action_rate_reward`, `joint_pos_limits_penalty`). You can adapt these concepts for your humanoid.
          * How to set up `CurriculumMgr` for velocity tracking.
          * How to set up randomization for terrain and external forces.
  * **`omni.isaac.orbit.tasks.manager_base.py` and `omni.isaac.orbit.tasks.rl_task.py`:** Understand how `_update_metrics` is used to log things.
  * **`omni.isaac.orbit.actuators.actuator_group`:** Learn about how to define different actuator types and apply commands.
  * **`omni.isaac.orbit.robots.articulated_robot`:** How to load and configure articulated robots.

**External Repos (less common but worth a search):**

  * Given how new Isaac Lab is, fully fleshed-out complex humanoid tasks might not be widely available as open-source projects outside of NVIDIA's examples yet. Most are still in internal development or academic research.
  * However, keep an eye on the NVIDIA Isaac Lab GitHub repository and related community forums (like the Isaac Sim Discord) for new examples or discussions as the ecosystem grows.

### Workflow Recommendation:

1.  **Start Simple:** Begin by just getting your humanoid loaded into Isaac Lab with basic controls. Focus on accurate URDF/USD loading.
2.  **Basic Task:** Implement a very simple `RLTask` with a minimal reward (e.g., just penalize falling).
3.  **Reward Components (One by One):** Gradually add your individual reward components (righting, kneeling, rising) and log them separately to ensure they behave as expected.
4.  **Action Rescaler:** Integrate your $\\beta$ action rescaler.
5.  **Curriculum:** Introduce curriculum learning, starting with the vertical force.
6.  **Domain Randomization:** Layer in initial posture variation and other domain randomizations.
7.  **Iterate and Debug:** RL development is highly iterative. Use the logging and visualization tools in Isaac Sim extensively.

Good luck with your porting effort, Francesca\! It sounds like a fantastic contribution to the Isaac Lab humanoid community.

**Comment by Coderxrohan** [comment]
Hi @francescacraievich,

That’s an exciting project. Porting from Legged-Gym to Isaac Lab (Orbit) is a non-trivial but very rewarding effort, especially given the structured approach of your algorithm. I’ve worked a bit with both, so I’ll try to point you in the right direction based on your questions:

1. Multi-Critic Reward Structure:

Isaac Orbit’s reward functions are modular and defined via RewardTerms. While multi-critic PPO isn’t built-in, you can simulate the structure by:

Splitting the reward into stage-specific components using custom flags or environment variables.

Logging these components separately (more on that below).

Training a unified policy but conditioning the reward functions dynamically per stage.

You could also explore using [RL Games](https://github.com/Denys88/rl_games) with Orbit, which supports custom reward shaping and has hooks you can leverage to implement multi-critic-like behavior.

2. Curriculum Learning:

Orbit provides CurriculumTerm support under RandomizationCfg, or you can implement curriculum via a callback that modifies environment parameters based on training step or agent success rate.

Example strategy:

Define curriculum variables in your environment config (like external forces).

Use a custom CurriculumManager or implement the logic in your reset() and step() hooks to update difficulty.

Also take a look at Orbit’s [FrankaCubeStack curriculum example](https://github.com/NVIDIA-Omniverse/Orbit/blob/main/source/standalone/environments/franka_cube_stack.py) for inspiration.

**Comment by github-actions** [comment]
🕒 **Discussion Activity Reminder** 🕒

This Discussion has been labeled as dormant by an automated system for having no activity in the last 60 days. Please consider one the following actions:

1️⃣ Close as Out of Date: If the topic is no longer relevant, close the Discussion as `out of date` at the bottom of the page.

2️⃣ Provide More Information: Share additional details or context — or let the community know if you've found a solution on your own.

3️⃣ Mark a Reply as Answer: If your question has been answered by a reply, mark the most helpful reply as the solution.

Note: This dormant notification will only apply to Discussions with the `Question` label. To learn more, see our [recent announcement](https://github.com/orgs/community/discussions/70478).

Thank you for helping bring this Discussion to a resolution! 💬
